lib/io.js

```diff
@@ -357,9 +353,10 @@ DatumReader.prototype = {
         var oldOffset = decoder._input.offset;
         var schemaIndex = decoder.readLong();
 
-        if (schemaIndex < 0 || schemaIndex >= writersSchema.schemas.length) {
```

Comment:
The issue with this conditional is that schemaIndex is a Long and not a JavaScript number.  readLong is indeed called externally and is expected to return a Long.  The whole reason for using Long is so we can handle data values that are full 64 bit.  That does mean that readInt will return a JavaScript number, but readLong returns an instance of Long.